test/hitl/card.binary/lib/notecard_binary/NotecardBinary.h

```diff
@@ -0,0 +1,420 @@
+        } else {
+            notecard.logDebugf("Successfully transferred %d(0x%x) bytes.\n", totalTransferSize, totalTransferSize);
+        }
+        if (!bytesRemaining) {
+            notecard.logDebug("Validating received binary data...\n");
+            if (!validateBinaryReceived(image, actualTransferSize, buffer, bufferLength)) {
+                return false;
+            }
+        }
+        NoteBinaryReset();
+        return bytesRemaining==0;
+    }
+
+    bool checkBufferSize(size_t chunkSize, size_t bufferSize)
+    {
+        size_t requiredSize = NoteBinaryRequiredTxBuffer(chunkSize);
+
+        NoteDebugf("Requested size: %u, available size: %u\n", requiredSize, bufferSize);
+
+        if (requiredSize > bufferSize) {
+            NoteDebugf("The required buffer size is larger than the allocated buffer: %d > %d\n", requiredSize, bufferSize);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @brief Compares the buffer sent to the notecard with the data read back.
+     *
+     * @param image The binary image to compare with the card.binary buffer on the Notecard.
+     * @param max_validation_length The maximum size to try to validate in one `card.binary.read` request.
+     * @param buffer Scratch buffer for running the validation.
+     * @param bufferLength Length of the scratch buffer.
+     * @return true
+     * @return false
+     */
+    bool validateBinaryReceived(BinaryGenerator& image, size_t expectedLength, uint8_t* buffer, size_t bufferLength)
+    {
+        size_t receivedLength = 0;
+        bool success = false;
+
+        image.reset();
+        size_t rxBufferSize = 0;
+        uint8_t* rxBuffer = nullptr;
+        const char* err;
+
+        if ((err = NoteBinaryRequiredRxBuffer(&rxBufferSize))!=nullptr) {
+            notecard.logDebugf("Error determining rx buffer size: %s\n", err);
+        } else if (nullptr==(rxBuffer=(uint8_t*)malloc(rxBufferSize))) {
+            notecard.logDebugf("Unable to allocate %d bytes for receive buffer.\n", rxBufferSize);
+        } else if ((err = NoteBinaryReceive(rxBuffer, rxBufferSize, &receivedLength))!=nullptr) {
+            notecard.logDebugf("Error receiving binary: %s\n", err);
+        } else if (expectedLength != receivedLength) {
+            notecard.logDebugf("Expected and actual received sizes differ: %d != %d\n", expectedLength, receivedLength);
+        } else {
+            notecard.logDebugf("Receive buffer size: %d\n", rxBufferSize);
+            // use the original buffer and chunk size to compare the buffers
+            size_t totalRead = 0;
+            while (totalRead < expectedLength) {
+                size_t thisChunk = std::min(bufferLength, expectedLength-image.offset());
+                size_t chunkRead = image.read(buffer, thisChunk);
+                assert(chunkRead == thisChunk);
+
+                for (size_t i=0; i<chunkRead; i++) {
```

Comment:
This would be especially interesting in the fail case, where the MD5 sum did not match. Although we don't have a guaranteed error message that would indicate this was the precise point of failure.

It would be cool to see how far it went off the rails and starting at which index.

Reply:
I'm not sure at present how we can fake a bad MD5 checksum. Perhaps by overriding the interface communications hooks passed to note-c so we can deliberately inject bad data?